src/box/applier.cc

```diff
@@ -1252,6 +1298,726 @@ applier_on_rollback(struct trigger *trigger, void *event)
 	return 0;
 }
 
+/** The underlying thread behind a number of appliers. */
+struct applier_thread {
+	/** A link in allocated threads list. */
+	struct rlist in_list;
+	struct cord cord;
+	/** The single thread endpoint. */
+	struct cbus_endpoint endpoint;
+	/** A pipe from the applier thread to tx. */
+	struct cpipe tx_pipe;
+	/** A pipe from tx to the applier thread. */
+	struct cpipe thread_pipe;
+	/** A conditional variable to wait on until thread initialization. */
+	struct fiber_cond ready_cond;
+	/**
+	 * A notification set for tx once the thread is operational: the
+	 * endpoint and both pipes are created.
+	 */
+	bool is_ready;
+};
+
+struct applier_msg {
+	struct cmsg base;
+	/** The function to execute in applier fiber. */
+	cmsg_f f;
+	/** The target applier this message is for. */
+	struct applier *applier;
+};
+
+struct applier_exit_msg {
+	struct applier_msg base;
+	/**
+	 * The thread's saved diag. To propagate exception to applier fiber in
+	 * tx thread.
+	 */
+	struct diag diag;
+};
+
+/**
+ * Propagate the exception causing the applier death to tx from applier
+ * thread.
+ */
+static void
+applier_exit_thread(struct cmsg *base)
+{
+	struct applier_exit_msg *msg = (struct applier_exit_msg *)base;
+	assert(!diag_is_empty(&msg->diag));
+	diag_move(&msg->diag, &fiber()->diag);
+	diag_raise();
+}
+
+struct applier_data_msg {
+	struct applier_msg base;
+	/** A list of read up transactions to be processed in tx thread. */
+	struct stailq txs;
+	/** This message's personal buffer, holding input rows. */
+	struct ibuf ibuf;
+	/** How much unparsed data there is in the ibuf. */
+	size_t parse_size;
+	/** This message's region used to allocate auxiliary structures. */
+	struct region region;
+};
+
+/* Applier thread related data. */
+struct applier_thread_data {
+	/** A pointer to the thread handling this applier's data stream. */
+	struct applier_thread *thread;
+	/**
+	 * A pair of rotating messages. While one of the messages is processed
+	 * in the tx thread the other is used to store incoming rows.
+	 */
+	struct applier_data_msg msgs[2];
+	/** An index of the message currently filling up in applier thread. */
+	int msg_ptr;
+	/**
+	 * A preallocated message used to notify tx that the applier should exit
+	 * due to a network error.
+	 */
+	struct applier_exit_msg exit_msg;
+	/** The reader fiber, reading and parsing incoming rows. */
+	struct fiber *reader;
+	/** A trigger invoked on reader fiber stop. */
+	struct trigger reader_on_stop;
+	/**
+	 * A conditional variable used by applier fiber in tx thread waiting for
+	 * the next message.
+	 */
+	struct fiber_cond msg_cond;
+	/**
+	 * A tx thread queue for messages delivered from applier thread but not
+	 * yet picked up by a corresponding applier.
+	 */
+	struct applier_msg *pending_msgs[2];
+	/** Pending message count. */
+	int pending_msg_cnt;
```

Comment:
It looks like this structure is modified by both applier and tx threads so it might be affected by false-sharing. I think you should clearly separate fields modified by different threads, like it's done in `iproto_connection`. Also, I'm not sure it's worth using a separate struct for this - it's name rather awkward. Maybe, just inline it in `struct applier`?